File: Data Structures & Algorithms/reverse-a-linked-list/submission-0.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */

class Solution {
public:
    ListNode* reverseList(ListNode* head) {
        ListNode* prev = nullptr;
        ListNode* curr = head;
        
        while (curr != nullptr) {
            ListNode* nxt = curr->next;  // 1. Save the next node
            curr->next = prev;           // 2. Reverse the link
            
            prev = curr;                 // 3. Move prev forward
            curr = nxt;                  // 4. Move curr forward
        }
        
        return prev;                     // prev is the new head
    }
};

File: Data Structures & Algorithms/linked-list-cycle-detection/submission-0.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def hasCycle(self, head: Optional[ListNode]) -> bool:
        visited = set()
        curr = head
        
        while curr:
            # If the current node object is already in our set, a cycle exists
            if curr in visited:
                return True
            
            # Add the current node reference to the set
            visited.add(curr)
            
            # Move to the next node
            curr = curr.next
            
        # If curr becomes None/null, we reached the end, so there is no cycle
        return False